File: apps/android/app/src/main/java/chat/rogi/rogichat/channelport/core/common/DonationFormatter.kt
package chat.rogi.rogichat.channelport.core.common

import java.text.NumberFormat
import java.util.Locale

object DonationFormatter {
    private val UNIT_BY_CURRENCY = mapOf(
        "SOOP_BALLOON" to "별풍선",
        "CHZZK_CHEESE" to "치즈",
        "CIME_BEAM" to "빔",
    )

    private val numberFormat: NumberFormat = NumberFormat.getNumberInstance(Locale.KOREA)

    /**
     * 후원 금액 표시 문자열 생성.
     * - native + 알려진 currency → "{N} {단위}"  예: "2 별풍선"
     * - krwSnapshot != null → "{N}원"
     * - 그 외 → 빈 문자열
     */
    fun format(
        nativeAmount: Int?,
        currency: String?,
        krwSnapshot: Int?,
    ): String {
        val unit = currency?.let { UNIT_BY_CURRENCY[it] }
        if (nativeAmount != null && unit != null) {
            return "${numberFormat.format(nativeAmount)} $unit"
        }
        if (krwSnapshot != null) {
            return "${numberFormat.format(krwSnapshot)}원"
        }
        return ""
    }
}

File: apps/ios/Sources/Features/Channel/DonationFormatter.swift
import Foundation

public enum DonationFormatter {
    /// 재화 키 → 표시 단위 매핑. 백엔드 PLATFORM_EXCHANGE_TABLE과 일치.
    private static let unitByCurrency: [String: String] = [
        "SOOP_BALLOON": "별풍선",
        "CHZZK_CHEESE": "치즈",
        "CIME_BEAM": "빔",
    ]

    /// 후원 금액 표시 문자열 생성.
    /// - native + 알려진 currency → "{N} {단위}"  (예: "2 별풍선")
    /// - 그 외 krwSnapshot ≠ nil → "{N}원"
    /// - 아무것도 없으면 빈 문자열
    public static func format(
        nativeAmount: Int?,
        currency: String?,
        krwSnapshot: Int?
    ) -> String {
        if let native = nativeAmount,
           let currencyKey = currency,
           let unit = unitByCurrency[currencyKey] {
            return "\(numberFormatter.string(from: NSNumber(value: native)) ?? "\(native)") \(unit)"
        }
        if let krw = krwSnapshot {
            return "\(numberFormatter.string(from: NSNumber(value: krw)) ?? "\(krw)")원"
        }
        return ""
    }

    private static let numberFormatter: NumberFormatter = {
        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.locale = Locale(identifier: "ko_KR")
        return f
    }()
}
